### PR TITLE
Issue #335: Checks under neverSuppressedChecks group should give more…

### PR DESCRIPTION
**Pull request by Brijeshthummar02**
Issue #335

Refines `neverSuppressedChecks` handling so violations are more precisely tied to patch-caused changes instead of always showing all violations in touched files.

## Changes

- Updated `SuppressionPatchFilterElement` to use causality-aware matching for never-suppressed checks.
- Added strategy-aware behavior (legacy behavior preserved for `newline` where causal attribution is ambiguous).
- Added precise handling for:
  - `UniqueProperties` (non-`newline`)
  - `RegexpMultiline` (non-`newline`)
- Updated `JavaPatchFilterElement` to make `CovariantEquals` more precise under non-`newline` strategies.
- Passed `strategy` into `SuppressionPatchFilterElement` from `SuppressionPatchFilter`.
- Updated expected outputs for `CovariantEquals` patchedline tests.

### CODE REVIEW

**Comment by romani** [comment]
We need more test to show clearly that it works.

**Comment by Brijeshthummar02** [comment]
> We need more test to show clearly that it works.

Sure, working on it.
Due to exams i am not able to contribute a lot and active. Will be back soon.

**Comment by Brijeshthummar02** [comment]
> We need more test to show clearly that it works.

In this PR, I added multiple `case2` scenarios specifically to make the behavior of `neverSuppressedChecks` more explicit and verifiable:

- Each new test file contains **multiple existing `CovariantEquals` violations**.
- The patch modifies **only one specific class scope**.
- The tests assert that **only the violation caused by the changed scope is reported**, while violations from **unchanged scopes are not reported**.
- These cases are implemented across both **`checkID` and `checkShortName`**, and for both **`context` and `patchedline`** modes to ensure consistent behavior.

The goal of these additions was to move beyond simple file-touch validation and instead **demonstrate causal precision**, making it clearer that `neverSuppressedChecks` behaves correctly in multi-violation scenarios.

If there are specific edge cases or additional scenarios you would like to see covered, I would be happy to extend the tests further.

**Comment by Brijeshthummar02** [comment]
> > We need more test to show clearly that it works.
> 
> In this PR, I added multiple `case2` scenarios specifically to make the behavior of `neverSuppressedChecks` more explicit and verifiable:
> 
> * Each new test file contains **multiple existing `CovariantEquals` violations**.
> * The patch modifies **only one specific class scope**.
> * The tests assert that **only the violation caused by the changed scope is reported**, while violations from **unchanged scopes are not reported**.
> * These cases are implemented across both **`checkID` and `checkShortName`**, and for both **`context` and `patchedline`** modes to ensure consistent behavior.
> 
> The goal of these additions was to move beyond simple file-touch validation and instead **demonstrate causal precision**, making it clearer that `neverSuppressedChecks` behaves correctly in multi-violation scenarios.
> 
> If there are specific edge cases or additional scenarios you would like to see covered, I would be happy to extend the tests further.



> > We need more test to show clearly that it works.
> 
> In this PR, I added multiple `case2` scenarios specifically to make the behavior of `neverSuppressedChecks` more explicit and verifiable:
> 
> * Each new test file contains **multiple existing `CovariantEquals` violations**.
> * The patch modifies **only one specific class scope**.
> * The tests assert that **only the violation caused by the changed scope is reported**, while violations from **unchanged scopes are not reported**.
> * These cases are implemented across both **`checkID` and `checkShortName`**, and for both **`context` and `patchedline`** modes to ensure consistent behavior.
> 
> The goal of these additions was to move beyond simple file-touch validation and instead **demonstrate causal precision**, making it clearer that `neverSuppressedChecks` behaves correctly in multi-violation scenarios.
> 
> If there are specific edge cases or additional scenarios you would like to see covered, I would be happy to extend the tests further.

@romani ping.

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java [hunk 1]
```diff
@@ -217,6 +221,53 @@ private boolean isNeverSuppressCheck(TreeWalkerAuditEvent event) {
...
+                }
+                else {
+                    result = true;
+                }
+            }
+        }
+        return result;
+    }
+
+    private boolean isChangedLineInsideEnclosingType(TreeWalkerAuditEvent event) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:279]: this method will always return true, as in java all files are types of any sort.

in what test it return false ?

I think you need to differentiate if violation happened in input file methods (return false) and on method (return true, as it like on type).

but on the same type  Check predicatably give violation on specific AST node, so we know by design of Check that it always give violaiton on methods 
https://checkstyle.org/checks/coding/covariantequals.html#Example1-code .

prove by test in what cases we need such   isChangedLineInsideEnclosingType, or remove it.

**Review by romani — CHANGES_REQUESTED** [review]
item:

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java (hunk 1 above)
Comment (Brijeshthummar02) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:279]: Thanks for the review, this is a valid concern.

`isChangedLineInsideEnclosingType(...)` is not always true.  
For `CovariantEquals`, violation is reported on method AST, but causality is type-level (missing `equals(Object)` in that class), so I attribute by enclosing type scope.

It returns `false` when no patch line is inside the violation’s enclosing type, proven by:
- `neversuppressedchecks/CovariantEquals/checkID/context/case2/defaultContextConfig.xml`
- `neversuppressedchecks/CovariantEquals/checkID/patchedline/case2/defaultContextConfig.xml`
- `neversuppressedchecks/CovariantEquals/checkShortName/context/case2/defaultContextConfig.xml`
- `neversuppressedchecks/CovariantEquals/checkShortName/patchedline/case2/defaultContextConfig.xml`

In all above, patch is empty and expected output is empty.

Also, precision improved in:
- `neversuppressedchecks/CovariantEquals/checkID/patchedline/expected.txt`
- `neversuppressedchecks/CovariantEquals/checkShortName/patchedline/expected.txt`

Now only line 11 is reported (line 4 is suppressed), showing we no longer report unchanged-type violations.

If you prefer method-level attribution instead of type-level attribution for this check, I can update implementation and tests accordingly.

Comment (Brijeshthummar02) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/JavaPatchFilterElement.java:279]: I agree that we should explicitly demonstrate when `isChangedLineInsideEnclosingType(...)` returns `false`.

I will reproduce this scenario locally and add a dedicated test case where a violation exists but no patch line falls inside the violation's enclosing type. This will help confirm whether the method is necessary or if the logic can be simplified.

I will update the PR once the test is added and verified.

**Comment by Brijeshthummar02** [comment]
@romani    I've pushed an update that explicitly covers the `isChangedLineInsideEnclosingType(...) == false` path.

In `case2` (for both check-id and check-short-name, in `context` and `patchedline`), I kept a real `CovariantEquals` violation in `Test`, but made the patch change only at a top-level line outside any type (`// patched line outside any type`).

`expected.txt` is empty in these `case2` scenarios, so this now shows: violation exists, file is touched, but no patch line falls inside the violation's enclosing type, and the violation is suppressed.

I also re-ran `mvn -Dtest=SuppressionJavaPatchFilterTest test` and `mvn clean verify` after the update, and it passes.

If you prefer, I can split this into a dedicated new test method name as well, but behavior-wise this scenario is now explicitly covered.

**Comment by Brijeshthummar02** [comment]
> @romani I've pushed an update that explicitly covers the `isChangedLineInsideEnclosingType(...) == false` path.
> 
> In `case2` (for both check-id and check-short-name, in `context` and `patchedline`), I kept a real `CovariantEquals` violation in `Test`, but made the patch change only at a top-level line outside any type (`// patched line outside any type`).
> 
> `expected.txt` is empty in these `case2` scenarios, so this now shows: violation exists, file is touched, but no patch line falls inside the violation's enclosing type, and the violation is suppressed.
> 
> I also re-ran `mvn -Dtest=SuppressionJavaPatchFilterTest test` and `mvn clean verify` after the update, and it passes.
> 
> If you prefer, I can split this into a dedicated new test method name as well, but behavior-wise this scenario is now explicitly covered.

@romani ping.

**Comment by romani** [comment]
@Brijeshthummar02  , please resole conflict

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilterElement.java [hunk 2]
```diff
@@ -154,10 +197,103 @@ private boolean isNeverSuppressCheck(AuditEvent event) {
...
+            if (!result) {
+                final String checkShortName = getCheckShortName(event)
+                        .replaceAll(CHECK_SUFFIX, "");
+                if (UNIQUE_PROPERTIES_CHECK.equals(checkShortName)
+                        && strategy != Strategy.NEWLINE) {
+                    result = isUniquePropertiesViolationCausedByPatch(event);
+                }
+                else if (REGEXP_MULTILINE_CHECK.equals(checkShortName)
+                        && strategy != Strategy.NEWLINE) {
+                    result = isRegexpMultilineViolationCausedByPatch(event);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilterElement.java:235]: this can cause all Check to be registered here ? 
it will be problem to maintain this, amount of Check is always grow, So this filter will have list of Checks it supports.

**Review by romani — CHANGES_REQUESTED** [review]
questions:

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilterElement.java (hunk 2 above)
Comment (Brijeshthummar02) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilterElement.java:235]: No, this won't require registering all checks here.

Most checks report violations on the same line, so the existing logic already works for them. Only a few checks (like `UniqueProperties` and `RegexpMultiline`) report violations on a different line than the actual change, which is why they need special handling.

For all other checks, the fallback still applies and shows violations in touched files, so nothing is missed.

Since only a very small number of checks behave this way, the scope remains limited and maintainable.

Comment (Brijeshthummar02) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilterElement.java:235]: `UniqueProperties` - reports on first duplicate occurrence, not necessarily the changed line\
`RegexpMultiline` - reports on first line of multi-line match, change may be on other lines.

Comment (Brijeshthummar02) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilterElement.java:235]: @romani if you prefer i can add comments about the same  at `isMatchingByNeverSuppressedCheck` method.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilterElement.java:235]: yes add, this methods should explain this tricky concept of such checks, there might be more Check like this discovered.

**Comment by romani** [comment]
conflict.

**Comment by Brijeshthummar02** [comment]
> conflict.

fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilterElement.java (hunk 2 above)
Comment (Brijeshthummar02) [reply src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressionPatchFilterElement.java:235]: Done, i have added clean comments in both the files. - `SuppressionPatchFilterElement.java` , `JavaPatchFilterElement.java`.

I think it makes much more sense now.

**Review by romani — APPROVED** [review]
thanks a lot
